File: src/dart/test/session_test.dart
import 'package:ores_dnd/ores_dnd.dart';
import 'package:test/test.dart';

import 'helpers.dart';

void main() {
  final traces = corpus()
      .where((c) => c.declaration == 'DndSessionTrace' && c.expectation == 'accepted')
      .map((c) => decodeDeclaration('DndSessionTrace', c.json) as DndSessionTrace)
      .toList();

  test('every shared DndSessionTrace replays identically', () {
    expect(traces.length, greaterThanOrEqualTo(20));
    final divergences = traces.map(replayTrace).whereType<TraceDivergence>().map((d) => d.toString()).toList();
    expect(divergences, isEmpty);
  });

  test('trace replay detects a divergence', () {
    final trace = traces.firstWhere((t) => t.id == 'basic-drop');
    final broken = DndSessionTrace(
      id: trace.id,
      inputs: trace.inputs,
      expected: [
        ...trace.expected.take(2),
        DndSessionSnapshot(state: DndSessionState.dropped, dragId: 'drag-0001', targetId: 'zone-a', operation: DndOperation.link),
      ],
    );
    expect(replayTrace(broken)?.step, 2);
  });

  test('session exposes envelope, result and listeners', () {
    final envelope = DndEnvelope.fromJson(readJson('instances/DndEnvelope/valid/text-copy.json'));
    final session = DndSession();
    final seen = <String>[];
    final unsubscribe = session.subscribe((snapshot, input) => seen.add('${input.kind.wire}:${snapshot.state.wire}'));
    expect(session.result, isNull);
    session.apply(DndSessionInput.start(envelope));
    expect(session.envelope?.dragId, envelope.dragId);
    session.apply(DndSessionInput.enter(const DndDropPolicy(targetId: 'zone-a', allowedOperations: [DndOperation.move], acceptedKinds: [DndItemKind.text])));
    expect(session.snapshot.isOverAcceptingTarget, isTrue);
    session.apply(const DndSessionInput.drop('zone-a'));
    expect(session.result, const DndDropResult(dragId: 'drag-0001', accepted: true, operation: DndOperation.move, targetId: 'zone-a'));
    unsubscribe();
    session.apply(const DndSessionInput.cancel());
    expect(seen, ['start:dragging', 'enter:over-target', 'drop:dropped']);
  });

  test('malformed inputs are ignored', () {
    final envelope = DndEnvelope.fromJson(readJson('instances/DndEnvelope/valid/text-copy.json'));
    final session = DndSession()..apply(DndSessionInput.start(envelope));
    final before = session.snapshot;
    expect(session.apply(const DndSessionInput(kind: DndSessionInputKind.enter, targetId: 'z')), before);
    expect(session.apply(const DndSessionInput(kind: DndSessionInputKind.leave)), before);
    expect(
      session.apply(DndSessionInput.enter(const DndDropPolicy(targetId: 'z', allowedOperations: [DndOperation.copy], acceptedKinds: [DndItemKind.text], maxItems: 0))),
      before,
    );
  });
}

File: src/ts/test/session.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { DndSession, inputs, replayTrace, resultOf } from "../dist/session.js";
import { traces, validEnvelope } from "./helpers.mjs";

test("every shared DndSessionTrace replays identically", async () => {
  const all = await traces();
  assert.ok(all.length >= 20, `expected the full corpus, got ${all.length}`);
  const divergences = all.map(replayTrace).filter(Boolean);
  assert.deepEqual(divergences, []);
});

test("trace replay detects a divergence", async () => {
  const [trace] = (await traces()).filter((t) => t.id === "basic-drop");
  trace.expected[2].operation = "link";
  const divergence = replayTrace(trace);
  assert.equal(divergence?.step, 2);
});

test("session exposes envelope, result and listeners", async () => {
  const envelope = await validEnvelope();
  const session = new DndSession();
  const seen = [];
  const unsubscribe = session.subscribe((snapshot, input) => seen.push(`${input.kind}:${snapshot.state}`));
  assert.equal(session.result, null);
  session.apply(inputs.start(envelope));
  assert.deepEqual(session.envelope, envelope);
  session.apply(inputs.enter({ targetId: "zone-a", allowedOperations: ["move"], acceptedKinds: ["text"] }));
  session.apply(inputs.drop("zone-a"));
  assert.deepEqual(session.result, { dragId: envelope.dragId, accepted: true, operation: "move", targetId: "zone-a" });
  unsubscribe();
  session.apply(inputs.cancel());
  assert.deepEqual(seen, ["start:dragging", "enter:over-target", "drop:dropped"]);
  assert.deepEqual(resultOf({ state: "cancelled", dragId: "d", errorCode: "cancelled" }), { dragId: "d", accepted: false, errorCode: "cancelled" });
  assert.equal(resultOf({ state: "dragging", dragId: "d" }), null);
});

test("malformed inputs are ignored", async () => {
  const envelope = await validEnvelope();
  const session = new DndSession();
  session.apply(inputs.start(envelope));
  const before = session.snapshot;
  assert.deepEqual(session.apply({ kind: "enter", targetId: "z" }), before);
  assert.deepEqual(session.apply({ kind: "leave" }), before);
  assert.deepEqual(session.apply({ kind: "enter", targetId: "z", policy: { targetId: "z", allowedOperations: ["copy"], acceptedKinds: ["text"], maxItems: 0 } }), before);
});
